src/connectivity/percolation.c

```diff
@@ -0,0 +1,339 @@
+ * reversed, it is the size of the largest component as vertices are removed
+ * in the reverse of the order given. If no vertex order is given, a random
+ * one will be used.
+ *
+ * \param graph The graph that vertices are assumed to be in. Edge directions
+ *    are ignored.
+ * \param output <code>output[i]</code> will contain the size of the largest
+ *    component after having added the vertex with index
+ *    <code>vertex_order[i]</code>.
+ * \param vertex_order The order the vertices are added in. Must not contain
+ *    duplicates. If \c NULL, a random order will be used.
+ * \return Error code.
+ *
+ * \sa \ref igraph_bond_percolation() to compute the edge percolation curve;
+ * \ref igraph_connected_components() to find the size of connected components.
+ *
+ * Time complexity: O(|V| + |E| a(|E|)) where a is the inverse Ackermann function,
+ * for all practical purposes it is not above 5.
+ */
+
+igraph_error_t igraph_site_percolation(
+        const igraph_t *graph,
+        igraph_vector_int_t *output,
+        const igraph_vector_int_t *vertex_order) {
+
+    const igraph_vector_int_t *p_vertex_order;
+    igraph_vector_int_t i_vertex_order;
+    const igraph_integer_t vcount = igraph_vcount(graph);
+
+    if (vertex_order == NULL) {
+        // Use random vertex order
+
+        IGRAPH_CHECK(igraph_vector_int_init_range(&i_vertex_order, 0, vcount));
+        IGRAPH_FINALLY(igraph_vector_int_destroy, &i_vertex_order);
+        IGRAPH_CHECK(igraph_vector_int_shuffle(&i_vertex_order));
+
+        p_vertex_order = &i_vertex_order;
+    } else {
+        p_vertex_order = vertex_order;
+    }
+
+    // Initialize variables
+    if (igraph_vector_int_size(p_vertex_order) != vcount) {
+        IGRAPH_ERRORF("Vertex order vector length (%" IGRAPH_PRId") "
+                      "does not match the vertex count (%" IGRAPH_PRId ").",
+                      IGRAPH_EINVAL,
+                      igraph_vector_int_size(p_vertex_order), vcount);
+    }
+    igraph_integer_t biggest = 1;
+
+    igraph_vector_int_t sizes;
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&sizes, vcount);
+
+    igraph_vector_int_t links;
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&links, vcount);
+
+    IGRAPH_CHECK(igraph_vector_int_resize(output, vcount));
+
+    for (igraph_integer_t i = 0; i < vcount; i++) {
+        VECTOR(sizes)[i] = 0;
+        VECTOR(links)[i] = i;
+    }
+
+    igraph_vector_int_t neighbors;
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&neighbors, 0);
+    for (igraph_integer_t i = 0; i < vcount; i++) {
+        IGRAPH_CHECK(percolate_site(graph, &links, &sizes, &biggest, VECTOR(*p_vertex_order)[i], &neighbors));
+        VECTOR(*output)[i] = biggest;
+    }
+    igraph_vector_int_destroy(&neighbors);
+    IGRAPH_FINALLY_CLEAN(1);
+
+    for (igraph_integer_t i = 0; i < vcount; i++) {
+        if (VECTOR(sizes)[i] == 0) {
+            IGRAPH_ERROR("Vertex order vector is missing vertices from graph.", IGRAPH_EINVAL);
+        }
+    }
```

Comment:
@Zepeacedust Isn't this check redundant with the check above on the size of `vertex_order`?